scripts/gen_packaging_stats.pl

```diff
@@ -0,0 +1,220 @@
+                            materials_parents => .. # stats for parents materials (e.g. PET will also count for plastic)
+                            materials => {
+                                "all" => ..
+                                "en:plastic" => 12, # number of products sold in France that are yogurts and that have a plastic bottle packaging component
+                            }
+                        },
+                        ..
+                    }
+                },
+                ..
+            }
+        },
+        ..
+    }
+}
+
+=cut
+
+use ProductOpener::PerlStandards;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Store qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Data qw/:all/;
+
+use File::Path qw(mkpath);
+use JSON::PP;
+use Data::DeepAccess qw(deep_exists deep_get deep_set deep_val);
+
+=head2 generate_packaging_stats_for_query($name, $query_ref)
+
+Generate packaging stats for products matching a specific query.
+
+Stats are saved in .sto format in $data_root/data/categories_stats/
+and in JSON format in $www_root/data/categories_stats/
+
+=head3 Arguments
+
+=head4 name $name
+
+
+
+=head4 query reference $query_ref
+
+=cut
+
+sub generate_packaging_stats_for_query ($name, $query_ref) {
+
+	$query_ref->{'empty'} = {"\$ne" => 1};
+	$query_ref->{'obsolete'} = {"\$ne" => 1};
+
+	my $fields_ref = {
+		countries_tags => 1,
+		categories_tags => 1,
+		packagings => 1,
+	};
+
+	my $socket_timeout_ms = 3 * 60 * 60 * 60000;    # 3 hours
+	my $products_collection = get_products_collection($socket_timeout_ms);
+
+	my $products_count = $products_collection->count_documents($query_ref);
+
+	print STDERR "$name: $products_count products\n";
+
+	my $cursor = $products_collection->query($query_ref)->sort({created_t => 1})->fields($fields_ref);
+
+	$cursor->immortal(1);
+
+	my $total = 0;
+
+	my $packagings_stats_ref = {};
+
+	# Go through all products
+	while (my $product_ref = $cursor->next) {
+		$total++;
+
+		if ($total % 1000 == 0) {
+			print STDERR "$name: $total / $products_count processed\n";
+		}
+
+		# Generate stats for all countries + en:world (products from all countries)
+		if (not defined $product_ref->{countries_tags}) {
+			$product_ref->{countries_tags} = [];
+		}
+		push @{$product_ref->{countries_tags}}, "en:world";
+
+		foreach my $country (@{$product_ref->{countries_tags}}) {
+
+			# Generate stats for all categories + all (products from all categories)
+			if (not defined $product_ref->{categories_tags}) {
+				$product_ref->{categories_tags} = [];
+			}
+			push @{$product_ref->{categories_tags}}, "all";
+
+			foreach my $category (@{$product_ref->{categories_tags}}) {
+
+				# Go through all packaging components
+				if (not defined $product_ref->{packagings}) {
+					$product_ref->{packagings} = [];
+				}
+
+				foreach my $packaging_ref (@{$product_ref->{packagings}}) {
```

Comment:
I would have change the order of those two loops for performance reason (avoid calling gen_tags_hierarchy_taxonomy for every country).

But it's not a big deal.

Reply:
Good point. I did it in this order because it was more logical in my head (I want packagings for all categories and for all countries), but indeed it multiplies the calls needlessly.